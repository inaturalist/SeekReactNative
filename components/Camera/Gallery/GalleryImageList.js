// @flow

import React, { useCallback, useContext, useEffect, useState } from "react";
import { Platform, FlatList } from "react-native";
import type { Node } from "react";
import { getPredictionsForImage } from "react-native-inat-camera";
import { useNavigation } from "@react-navigation/native";

import { checkForPhotoMetaData } from "../../../utility/photoHelpers";
import { viewStyles } from "../../../styles/camera/gallery";
import { dirTaxonomy, dirModel } from "../../../utility/dirStorage";
import { ObservationContext, UserContext } from "../../UserContext";
import { dimensions } from "../../../styles/global";
import GalleryImage from "./GalleryImage";

type Props = {
  photos: Array<Object>,
  onEndReached: Function,
  setLoading: ( ) => void
}

const GalleryImageList = ( { onEndReached, photos, setLoading }: Props ): Node => {
  const { setObservation, observation } = useContext( ObservationContext );
  const { login } = useContext( UserContext );
  const navigation = useNavigation( );
  const [imageSelected, setImageSelected] = useState( false );

  const navigateToResults = useCallback( ( uri, time, location, predictions ) => {
    const { navigate } = navigation;

    const image = {
      time,
      uri,
      predictions: [],
      errorCode: 0,
      latitude: null,
      longitude: null,
      preciseCoords: {},
      onlineVision: false
    };

    if ( checkForPhotoMetaData( location ) ) {
      const { latitude, longitude } = location;
      image.latitude = latitude || null;
      image.longitude = longitude || null;

      if ( login ) {
        image.preciseCoords = {
          latitude,
          longitude,
          accuracy: null
        };
      }
    } else if ( login ) {
      image.preciseCoords = {
        latitude: null,
        longitude: null,
        accuracy: null
      };
    }

    if ( predictions && predictions.length > 0 ) {
      image.predictions = predictions;
      setObservation( { image } );
    } else {
      image.onlineVision = true;
      setObservation( { image } );
      navigate( "Confirm" );
    }
  }, [navigation, setObservation, login] );

  useEffect( ( ) => {
    if ( observation
      && observation.taxon
      && !observation.image.onlineVision
      && imageSelected
    ) {
      // changed to navigate from push bc on Android, with RN > 0.65.x, the camera was
      // popping up over the top of the match screen
      navigation.navigate( "Drawer", {
        screen: "Match"
      } );
    }
  }, [observation, navigation, imageSelected] );

  const getPredictions = useCallback( ( uri, timestamp, location ) => {
    const path = uri.split( "file://" );
    const reactUri = path[1];

    getPredictionsForImage( {
      uri: reactUri,
      modelFilename: dirModel,
      taxonomyFilename: dirTaxonomy
    } ).then( ( { predictions } ) => {
      navigateToResults( uri, timestamp, location, predictions );
    } ).catch( ( err ) => {
      console.log( "Error", err );
    } );
  }, [navigateToResults] );

  const selectImage = useCallback( ( item ) => {
    setImageSelected( true );
    setLoading( );
    const { timestamp, location, image } = item.node;

    if ( Platform.OS === "android" ) {
      getPredictions( image.uri, timestamp, location );
    } else {
      navigateToResults( image.uri, timestamp, location );
    }
  }, [getPredictions, navigateToResults, setLoading] );

  const renderImage = useCallback( ( { item } ) => <GalleryImage item={item} selectImage={selectImage} />, [selectImage] );

  // skips measurement of dynamic content for faster loading
  const getItemLayout = useCallback( ( data, index ) => ( {
    length: ( dimensions.width / 4 - 2 ),
    offset: ( dimensions.width / 4 - 2 ) * index,
    index
  } ), [] );

  const extractKey = useCallback( ( item, index ) => `${item}${index}`, [] );

  return (
    <FlatList
      testID="gallery-image-list"
      // estimatedItemSize={97}
      data={photos}
      contentContainerStyle={viewStyles.grayContainer}
      getItemLayout={getItemLayout}
      initialNumToRender={4}
      keyExtractor={extractKey}
      numColumns={4}
      onEndReachedThreshold={0.2}
      onEndReached={onEndReached}
      renderItem={renderImage}
    />
  );
};

export default GalleryImageList;
