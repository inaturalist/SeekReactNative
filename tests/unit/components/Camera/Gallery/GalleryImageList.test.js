import React from "react";
import { render, screen } from "tests/jest-utils";

import GalleryImageList from "../../../../../components/Camera/Gallery/GalleryImageList";

const mockPhotos = [
  {
    node: {
      group_name: "All Photos",
      image: {
        fileSize: null,
        filename: null,
        height: null,
        playableDuration: null,
        uri: "some_uri_1",
        width: null
      },
      location: null,
      timestamp: 1669976394.6891332,
      type: "image"
    }
  },
  {
    node: {
      group_name: "All Photos",
      image: {
        fileSize: null,
        filename: null,
        height: null,
        playableDuration: null,
        uri: "some_uri_2",
        width: null
      },
      location: null,
      timestamp: 1669976352.0699105,
      type: "image"
    }
  },
  {
    node: {
      group_name: "All Photos",
      image: {
        fileSize: null,
        filename: null,
        height: null,
        playableDuration: null,
        uri: "some_uri_3",
        width: null
      },
      location: null,
      timestamp: 1669976241.7418242,
      type: "image"
    }
  },
  {
    node: {
      group_name: "All Photos",
      image: {
        fileSize: null,
        filename: null,
        height: null,
        playableDuration: null,
        uri: "some_uri_4",
        width: null
      },
      location: null,
      timestamp: 1669801317.2202396,
      type: "image"
    }
  },
  {
    node: {
      group_name: "All Photos",
      image: {
        fileSize: null,
        filename: null,
        height: null,
        playableDuration: null,
        uri: "some_uri_5",
        width: null
      },
      location: null,
      timestamp: 1669801317.2202396,
      type: "image"
    }
  }
];

const containerID = "gallery-image-list";

describe( "GalleryImageList", () => {
  test( "should render correctly", async () => {
    render( <GalleryImageList onEndReached={() => jest.fn()} photos={mockPhotos} setLoading={() => jest.fn()} /> );
    const container = await screen.findByTestId( containerID );
    expect( container ).toBeTruthy();
    expect( screen ).toMatchSnapshot();
  } );
} );
