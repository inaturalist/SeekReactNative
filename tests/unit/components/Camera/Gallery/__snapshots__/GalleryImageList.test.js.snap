// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryImageList should render correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "backgroundColor": "#f5f5f5",
    }
  }
  data={
    [
      {
        "node": {
          "group_name": "All Photos",
          "image": {
            "fileSize": null,
            "filename": null,
            "height": null,
            "playableDuration": null,
            "uri": "some_uri_1",
            "width": null,
          },
          "location": null,
          "timestamp": 1669976394.6891332,
          "type": "image",
        },
      },
      {
        "node": {
          "group_name": "All Photos",
          "image": {
            "fileSize": null,
            "filename": null,
            "height": null,
            "playableDuration": null,
            "uri": "some_uri_2",
            "width": null,
          },
          "location": null,
          "timestamp": 1669976352.0699105,
          "type": "image",
        },
      },
      {
        "node": {
          "group_name": "All Photos",
          "image": {
            "fileSize": null,
            "filename": null,
            "height": null,
            "playableDuration": null,
            "uri": "some_uri_3",
            "width": null,
          },
          "location": null,
          "timestamp": 1669976241.7418242,
          "type": "image",
        },
      },
      {
        "node": {
          "group_name": "All Photos",
          "image": {
            "fileSize": null,
            "filename": null,
            "height": null,
            "playableDuration": null,
            "uri": "some_uri_4",
            "width": null,
          },
          "location": null,
          "timestamp": 1669801317.2202396,
          "type": "image",
        },
      },
      {
        "node": {
          "group_name": "All Photos",
          "image": {
            "fileSize": null,
            "filename": null,
            "height": null,
            "playableDuration": null,
            "uri": "some_uri_5",
            "width": null,
          },
          "location": null,
          "timestamp": 1669801317.2202396,
          "type": "image",
        },
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  initialNumToRender={4}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.2}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  testID="gallery-image-list"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      style={null}
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityLabel="some_uri_1"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 1,
              "paddingTop": 2,
            }
          }
        >
          <Image
            source={
              {
                "uri": "some_uri_1",
              }
            }
            style={
              {
                "height": 185.5,
                "width": 185.5,
              }
            }
          />
        </View>
        <View
          accessibilityLabel="some_uri_2"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 1,
              "paddingTop": 2,
            }
          }
        >
          <Image
            source={
              {
                "uri": "some_uri_2",
              }
            }
            style={
              {
                "height": 185.5,
                "width": 185.5,
              }
            }
          />
        </View>
        <View
          accessibilityLabel="some_uri_3"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 1,
              "paddingTop": 2,
            }
          }
        >
          <Image
            source={
              {
                "uri": "some_uri_3",
              }
            }
            style={
              {
                "height": 185.5,
                "width": 185.5,
              }
            }
          />
        </View>
        <View
          accessibilityLabel="some_uri_4"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 1,
              "paddingTop": 2,
            }
          }
        >
          <Image
            source={
              {
                "uri": "some_uri_4",
              }
            }
            style={
              {
                "height": 185.5,
                "width": 185.5,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={null}
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityLabel="some_uri_5"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 1,
              "paddingTop": 2,
            }
          }
        >
          <Image
            source={
              {
                "uri": "some_uri_5",
              }
            }
            style={
              {
                "height": 185.5,
                "width": 185.5,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
